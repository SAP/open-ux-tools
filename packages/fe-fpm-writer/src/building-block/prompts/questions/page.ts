import type { Answers } from 'inquirer';
import { i18nNamespaces, translate } from '../../../i18n';
import { getBuildingBlockIdPrompt, getViewOrFragmentPathPrompt, getAggregationPathPrompt } from '../utils';
import type { PromptContext, Prompts } from '../../../prompts/types';
import { BuildingBlockType } from '../../types';
import type { BuildingBlockConfig, Page } from '../../types';
import { SapShortTextType, SapLongTextType } from '@sap-ux/i18n';

export type PagePromptsAnswer = BuildingBlockConfig<Page> & Answers;

/**
 * Returns a list of prompts required to generate a page building block.
 *
 * @param context - prompt context including data about project
 * @returns Prompt with questions for page.
 */
export async function getPageBuildingBlockPrompts(context: PromptContext): Promise<Prompts<PagePromptsAnswer>> {
    const t = translate(i18nNamespaces.buildingBlock, 'prompts.page.');

    return {
        questions: [
            getViewOrFragmentPathPrompt(context, t('viewOrFragmentPath.validate') as string, {
                message: t('viewOrFragmentPath.message') as string,
                guiOptions: {
                    mandatory: true,
                    dependantPromptNames: ['aggregationPath']
                }
            }),
            getAggregationPathPrompt(context, {
                message: t('aggregation') as string,
                guiOptions: {
                    mandatory: true
                }
            }),
            getBuildingBlockIdPrompt(context, t('id.validation') as string, {
                message: t('id.message') as string,
                default: 'Page',
                guiOptions: {
                    mandatory: true
                }
            }),
            {
                type: 'input',
                name: 'buildingBlockData.title',
                message: t('title.message') as string,
                guiOptions: {
                    mandatory: true,
                    translationProperties: {
                        type: SapShortTextType.Heading,
                        annotation: t('title.translationAnnotation') as string
                    }
                }
            },
            {
                type: 'input',
                name: 'buildingBlockData.description',
                message: t('description.message') as string,
                guiOptions: {
                    translationProperties: {
                        type: SapLongTextType.Description,
                        annotation: t('description.translationAnnotation') as string
                    }
                }
            }
        ],
        initialAnswers: {
            buildingBlockData: {
                buildingBlockType: BuildingBlockType.Page
            }
        }
    };
}
