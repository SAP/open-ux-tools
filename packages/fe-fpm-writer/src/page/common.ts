import type { Editor } from 'mem-fs-editor';
import { join } from 'path';
import type { ManifestNamespace } from '@sap-ux/project-access';
import { validateBasePath } from '../common/validate';
import type {
    CustomPage,
    FCL,
    InternalCustomPage,
    InternalObjectPage,
    ObjectPage,
    Navigation,
    InternalListReport
} from './types';
import type { Manifest } from '../common/types';
import { FCL_ROUTER } from '../common/defaults';

/**
 * Suffix for patterns to support arbitrary paramters
 */
export const PATTERN_SUFFIX = ':?query:';

/**
 * Generates the pattern for a new route based on the input.
 *
 * @param routes existing routes
 * @param targetEntity entity of the target
 * @param nav navigation object
 * @returns the generated pattern as string
 */
export function generateRoutePattern(
    routes: ManifestNamespace.Route[],
    targetEntity: string,
    nav?: Navigation
): string {
    const parts: string[] = [];
    if (nav?.sourcePage && nav.navEntity) {
        const sourceRoute = routes.find((route) => route.name === nav.sourcePage);
        if (sourceRoute?.pattern) {
            const basePattern = sourceRoute.pattern.replace(PATTERN_SUFFIX, '');
            if (basePattern) {
                parts.push(basePattern);
                parts.push('/');
                parts.push(nav.navEntity);
            } else {
                parts.push(targetEntity);
            }
        } else {
            throw new Error('Navigation source invalid');
        }
    } else if (routes.length > 0) {
        parts.push(targetEntity);
    }

    if (nav?.navKey) {
        parts.push(`({${nav?.navEntity ?? targetEntity}Key})`);
    }
    parts.push(PATTERN_SUFFIX);
    return parts.join('');
}

/**
 * Generates the target property for a route based on exiting routes, configurations and the target's name.
 *
 * @param routes existing routes
 * @param name name of the target page
 * @param fcl optional flag if FCL is enabled
 * @param nav navigation object
 * @returns the target property of a route
 */
export function generateRouteTarget(
    routes: ManifestNamespace.Route[],
    name: string,
    fcl?: boolean,
    nav?: Navigation
): string | [string] | [ManifestNamespace.RouteTargetObject] {
    if (nav?.sourcePage) {
        const sourceRoute = routes.find((route) => route.name === nav.sourcePage);
        // FCL only supports 3 columns, therefore, show the page in fullscreen if it is the 4th level of navigation
        if (fcl && sourceRoute?.target?.constructor === Array && sourceRoute.target.length < 3) {
            return [...sourceRoute.target, name] as [string];
        }
    }
    return fcl ? [name] : name;
}

/**
 * Create a function that can be used as JsonReplace when calling extendJson.
 *
 * @param config page configuration
 * @returns a JsonReplacer function for the usage in ejs
 */
export function getManifestJsonExtensionHelper(
    config: InternalCustomPage | InternalObjectPage | InternalListReport
): (key: string, value: any) => any {
    return (key, value) => {
        switch (key) {
            case 'routing':
                value.routes = value.routes ?? [];
                break;
            case 'routes':
                const routes = value as ManifestNamespace.Route[];
                routes.push({
                    name: `${config.entity}${config.name}`,
                    pattern: generateRoutePattern(routes, config.entity, config.navigation),
                    target: generateRouteTarget(routes, `${config.entity}${config.name}`, config.fcl, config.navigation)
                });
                break;
            default:
                break;
        }
        return value;
    };
}

/**
 * Get the configuration parameters for the flexible column layout based on the given manifest and navigation config.
 *
 * @param manifest existing manifest
 * @param navigation navigation configuration that is to be added
 * @returns fcl configuration
 */
export function getFclConfig(manifest: Manifest, navigation?: Navigation): FCL {
    const config: FCL = {};
    if (manifest['sap.ui5']?.routing?.config?.routerClass === FCL_ROUTER) {
        config.fcl = true;
        if (navigation) {
            const sourceRoute = ((manifest['sap.ui5']?.routing?.routes as ManifestNamespace.Route[]) || []).find(
                (route) => route.name === navigation?.sourcePage
            );
            config.controlAggregation =
                ((sourceRoute?.target as string[]) ?? []).length > 1 ? 'endColumnPages' : 'midColumnPages';
        } else {
            config.controlAggregation = 'beginColumnPages';
        }
    }
    return config;
}

/**
 * Validate the input parameters for the generation of a custom or an object page.
 *
 * @param basePath - the base path
 * @param config - the custom page configuration
 * @param fs - the memfs editor instance
 * @returns the updated memfs editor instance
 */
export function validatePageConfig(basePath: string, config: CustomPage | ObjectPage, fs: Editor): Editor {
    // common validators

    validateBasePath(basePath, fs);

    // validate config against the manifest
    if (config.navigation?.sourcePage) {
        const manifest = fs.readJSON(join(basePath, 'webapp/manifest.json')) as Manifest;
        if (!manifest['sap.ui5']?.routing?.targets?.[config.navigation.sourcePage]) {
            throw new Error(`Could not find navigation source ${config.navigation.sourcePage}!`);
        }
        const routes: { [k: string]: ManifestNamespace.RouteWithoutName } = {};
        if (manifest['sap.ui5']?.routing?.routes?.constructor === Array) {
            manifest['sap.ui5'].routing.routes.forEach((routeWithName) => {
                routes[routeWithName.name] = routeWithName;
            });
        } else {
            Object.assign(routes, manifest['sap.ui5']?.routing?.routes ?? {});
        }

        const route = routes[config.navigation.sourcePage];
        if (!route || !route.pattern || !route.target) {
            throw new Error(`Invalid routing configuration for navigation source ${config.navigation.sourcePage}!`);
        }
    }

    return fs;
}
