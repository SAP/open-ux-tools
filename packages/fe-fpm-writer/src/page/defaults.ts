import { Editor } from 'mem-fs-editor';

import { CustomPage, InternalCustomPage } from './types';
import { setCommonDefaults } from '../common/defaults';

/**
 * Enhances the provided custom page configuration with default data.
 *
 * @param {CustomPage} data - a custom page configuration object
 * @param manifestPath - path to the application manifest
 * @param fs - mem-fs reference to be used for file access
 * @returns enhanced configuration
 */
export function enhanceData(data: CustomPage, manifestPath: string, fs: Editor): InternalCustomPage {
    const manifest: any = fs.readJSON(manifestPath);

    // set common defaults
    const config = setCommonDefaults<CustomPage>(data, manifestPath, manifest);

    if (config.view === undefined) {
        config.view = {
            title: config.name
        };
    }
    return config;
}
