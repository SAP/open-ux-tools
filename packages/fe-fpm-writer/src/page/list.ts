import { join } from 'path';
import { create as createStorage } from 'mem-fs';
import type { Editor } from 'mem-fs-editor';
import { create } from 'mem-fs-editor';
import { render } from 'ejs';
import { getFclConfig, getManifestJsonExtensionHelper, validatePageConfig } from './common';
import type { Manifest } from '../common/types';
import type { ListReport, InternalListReport } from './types';
import { getTemplatePath } from '../templates';

/**
 * Enhances the provided list report configuration with default data.
 *
 * @param data - a list report configuration object
 * @param manifest - application manifest
 * @returns enhanced configuration
 */
function enhanceData(data: ListReport, manifest: Manifest): InternalListReport {
    const config: InternalListReport = { settings: {}, ...data, name: 'ListReport', ...getFclConfig(manifest) };

    // move settings into correct possition in the manifest
    config.settings.entitySet = data.entity;
    config.settings.navigation = {};
    // use standard file name if i18n enhancement required
    if (config.settings.enhanceI18n === true) {
        config.settings.enhanceI18n = `i18n/custom${config.entity}${config.name}.properties`;
    }
    // move table settings into the correct structure
    if (config.settings.tableSettings) {
        config.settings.controlConfiguration = config.settings.controlConfiguration ?? {};
        config.settings.controlConfiguration['@com.sap.vocabularies.UI.v1.LineItem'] =
            config.settings.controlConfiguration['@com.sap.vocabularies.UI.v1.LineItem'] ?? {};
        config.settings.controlConfiguration['@com.sap.vocabularies.UI.v1.LineItem'].tableSettings =
            config.settings.tableSettings;
        delete config.settings.tableSettings;
    }
    return config;
}

/**
 * Add a ListReport to an existing UI5 application.
 *
 * @param basePath - the base path
 * @param data - the object page configuration
 * @param fs - the memfs editor instance
 * @returns the updated memfs editor instance
 */
export function generate(basePath: string, data: ListReport, fs?: Editor): Editor {
    if (!fs) {
        fs = create(createStorage());
    }
    validatePageConfig(basePath, data, fs);

    const manifestPath = join(basePath, 'webapp/manifest.json');
    const manifest = fs.readJSON(manifestPath) as Manifest;

    const config = enhanceData(data, manifest);

    // enhance manifest.json
    fs.extendJSON(
        manifestPath,
        JSON.parse(render(fs.read(getTemplatePath('page/list/manifest.json')), config, {})),
        getManifestJsonExtensionHelper(config)
    );

    return fs;
}
