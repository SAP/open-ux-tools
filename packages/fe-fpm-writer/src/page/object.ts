import { join } from 'path';
import { create as createStorage } from 'mem-fs';
import type { Editor } from 'mem-fs-editor';
import { create } from 'mem-fs-editor';
import { render } from 'ejs';
import { getFclConfig, getManifestJsonExtensionHelper, validatePageConfig } from './common';
import type { Manifest } from '../common/types';
import type { ObjectPage, InternalObjectPage } from './types';
import { getTemplatePath } from '../templates';

/**
 * Enhances the provided list report configuration with default data.
 *
 * @param data - a list report configuration object
 * @param manifest - application manifest
 * @returns enhanced configuration
 */
function enhanceData(data: ObjectPage, manifest: Manifest): InternalObjectPage {
    const config: InternalObjectPage = { settings: {}, ...data, name: 'ObjectPage', ...getFclConfig(manifest) };

    // move settings into correct possition in the manifest
    config.settings.entitySet = data.entity;
    config.settings.navigation = {};
    // use standard file name if i18n enhancement required
    if (config.settings.enhanceI18n === true) {
        config.settings.enhanceI18n = `i18n/custom${config.entity}${config.name}.properties`;
    }
    return config;
}

/**
 * Add an object page to an existing UI5 application.
 *
 * @param basePath - the base path
 * @param data - the object page configuration
 * @param fs - the memfs editor instance
 * @returns the updated memfs editor instance
 */
export function generate(basePath: string, data: ObjectPage, fs?: Editor): Editor {
    if (!fs) {
        fs = create(createStorage());
    }
    validatePageConfig(basePath, data, fs);

    const manifestPath = join(basePath, 'webapp/manifest.json');
    const manifest = fs.readJSON(manifestPath) as Manifest;

    const config = enhanceData(data, manifest);

    // enhance manifest.json
    fs.extendJSON(
        manifestPath,
        JSON.parse(render(fs.read(getTemplatePath('/page/object/manifest.json')), config, {})),
        getManifestJsonExtensionHelper(config)
    );

    return fs;
}
