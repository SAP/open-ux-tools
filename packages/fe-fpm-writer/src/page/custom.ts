import { join } from 'path';
import { create as createStorage } from 'mem-fs';
import type { Editor } from 'mem-fs-editor';
import { create } from 'mem-fs-editor';
import { render } from 'ejs';
import type { CustomPage, InternalCustomPage } from './types';
import { getFclConfig, getManifestJsonExtensionHelper, validatePageConfig } from './common';
import { setCommonDefaults } from '../common/defaults';
import type { Manifest } from '../common/types';
import { validateVersion } from '../common/validate';
import { getTemplatePath } from '../templates';
import { coerce } from 'semver';
import { addExtensionTypes } from '../common/utils';

/**
 * Enhances the provided custom page configuration with default data.
 *
 * @param data - a custom page configuration object
 * @param manifestPath - path to the application manifest
 * @param fs - mem-fs reference to be used for file access
 * @returns enhanced configuration
 */
export function enhanceData(data: CustomPage, manifestPath: string, fs: Editor): InternalCustomPage {
    const manifest = fs.readJSON(manifestPath) as Manifest;

    // set common defaults
    const config = setCommonDefaults(data, manifestPath, manifest) as InternalCustomPage;

    // set FCL configuration
    const fclConfig = getFclConfig(manifest, config.navigation);
    config.fcl = fclConfig.fcl;
    config.controlAggregation = fclConfig.controlAggregation;

    if (config.view === undefined) {
        config.view = {
            title: config.name
        };
    }
    return config;
}

/**
 * Validate the UI5 version and if valid return the root folder for the templates to be used.
 *
 * @param ui5Version - optional minimum required UI5 version
 * @returns root folder  containg the templates if the version is supported otherwise throws an error
 */
export function getTemplateRoot(ui5Version?: string): string {
    const minVersion = coerce(ui5Version);
    if (!minVersion || minVersion.minor >= 94) {
        return getTemplatePath('/page/custom/1.94');
    } else {
        return getTemplatePath('/page/custom/1.84');
    }
}

/**
 * Add a custom page to an existing UI5 application.
 *
 * @param {string} basePath - the base path
 * @param {CustomPage} data - the custom page configuration
 * @param {Editor} [fs] - the memfs editor instance
 * @returns {Promise<Editor>} the updated memfs editor instance
 */
export function generate(basePath: string, data: CustomPage, fs?: Editor): Editor {
    if (!fs) {
        fs = create(createStorage());
    }
    validateVersion(data.minUI5Version);
    validatePageConfig(basePath, data, fs);

    const manifestPath = join(basePath, 'webapp/manifest.json');

    const config = enhanceData(data, manifestPath, fs);

    // merge content into existing files
    const root = getTemplateRoot(data.minUI5Version);

    // enhance manifest.json
    fs.extendJSON(
        manifestPath,
        JSON.parse(render(fs.read(join(root, `manifest.json`)), config, {})),
        getManifestJsonExtensionHelper(config)
    );

    // add extension content
    const viewPath = join(config.path, `${config.name}.view.xml`);
    if (!fs.exists(viewPath)) {
        fs.copyTpl(join(root, 'ext/View.xml'), viewPath, config);
    }

    const ext = data.typescript ? 'ts' : 'js';
    const controllerPath = join(config.path, `${config.name}.controller.${ext}`);
    if (!fs.exists(controllerPath)) {
        fs.copyTpl(join(root, `ext/Controller.${ext}`), controllerPath, config);
    }

    if (data.typescript) {
        addExtensionTypes(basePath, data.minUI5Version, fs);
    }

    return fs;
}
