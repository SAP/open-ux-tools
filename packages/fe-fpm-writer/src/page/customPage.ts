import { join } from 'path';
import { create as createStorage } from 'mem-fs';
import type { Editor } from 'mem-fs-editor';
import { create } from 'mem-fs-editor';
import { render } from 'ejs';
import { enhanceData } from './defaults';
import type { CustomPage } from './types';
import { validateVersion } from '../common/validate';
import { getManifestJsonExtensionHelper, validatePageConfig } from './common';

/**
 * Validate the UI5 version and if valid return the root folder for the templates to be used.
 *
 * @param ui5Version - optional minimum required UI5 version
 * @returns root folder  containg the templates if the version is supported otherwise throws an error
 */
export function getTemplateRoot(ui5Version?: number): string {
    if (ui5Version === undefined || ui5Version >= 1.94) {
        return join(__dirname, '../../templates/page/1.94');
    } else {
        return join(__dirname, '../../templates/page/1.84');
    }
}

/**
 * Add a custom page to an existing UI5 application.
 *
 * @param {string} basePath - the base path
 * @param {CustomPage} data - the custom page configuration
 * @param {Editor} [fs] - the memfs editor instance
 * @returns {Promise<Editor>} the updated memfs editor instance
 */
export function generate(basePath: string, data: CustomPage, fs?: Editor): Editor {
    if (!fs) {
        fs = create(createStorage());
    }
    validateVersion(data.ui5Version);
    validatePageConfig(basePath, data, fs);

    const manifestPath = join(basePath, 'webapp/manifest.json');
    const config = enhanceData(data, manifestPath, fs);

    // merge content into existing files
    const root = getTemplateRoot(data.ui5Version);

    // enhance manifest.json
    fs.extendJSON(
        manifestPath,
        JSON.parse(render(fs.read(join(root, `manifest.json`)), config)),
        getManifestJsonExtensionHelper(config)
    );

    // add extension content
    const viewPath = join(config.path, `${config.name}.view.xml`);
    if (!fs.exists(viewPath)) {
        fs.copyTpl(join(root, 'ext/View.xml'), viewPath, config);
    }
    const controllerPath = join(config.path, `${config.name}.controller.js`);
    if (!fs.exists(controllerPath)) {
        fs.copyTpl(join(root, 'ext/Controller.js'), controllerPath, config);
    }

    return fs;
}
