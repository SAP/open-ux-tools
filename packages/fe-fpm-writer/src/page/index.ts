import type { CustomPage, ObjectPage, ListReport, Navigation } from './types';
import { validatePageConfig } from './common';
import { generate as generateCustomPage } from './custom';
import { generate as generateObjectPage } from './object';
import { generate as generateListReport } from './list';

export {
    validatePageConfig,
    generateCustomPage,
    generateObjectPage,
    generateListReport,
    CustomPage,
    ObjectPage,
    ListReport,
    Navigation
};
