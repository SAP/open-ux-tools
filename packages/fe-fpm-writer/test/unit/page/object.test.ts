import { create as createStorage } from 'mem-fs';
import type { Editor } from 'mem-fs-editor';
import { create } from 'mem-fs-editor';
import { join } from 'path';
import type { ManifestNamespace } from '@sap-ux/project-access';
import type { ObjectPage } from '../../../src/page';
import { generate } from '../../../src/page/object';

describe('ObjectPage', () => {
    const testDir = '' + Date.now();
    let fs: Editor;

    const testAppManifest = JSON.stringify(
        {
            'sap.app': {
                id: 'my.test.App'
            },
            'sap.ui5': {
                dependencies: {
                    libs: {
                        'sap.fe.templates': {}
                    }
                },
                routing: {
                    routes: [
                        {
                            pattern: ':?query:',
                            name: 'RootEntityListReport',
                            target: 'TestListReport'
                        },
                        {
                            pattern: 'RootEntity({RootEntityKey}):?query:',
                            name: 'RootEntityObjectPage',
                            target: 'RootEntityObjectPage'
                        }
                    ] as ManifestNamespace.Route[],
                    targets: {
                        RootEntityListReport: {},
                        RootEntityObjectPage: {}
                    }
                }
            }
        },
        null,
        2
    );

    beforeEach(() => {
        fs = create(createStorage());
        fs.delete(testDir);
    });

    describe('generate', () => {
        const minimalInput: ObjectPage = {
            entity: 'OtherEntity'
        };

        test('minimal input', () => {
            const target = join(testDir, 'minimal-input');
            fs.write(join(target, 'webapp/manifest.json'), testAppManifest);
            generate(target, minimalInput, fs);

            expect(fs.readJSON(join(target, 'webapp/manifest.json'))).toMatchSnapshot();
        });

        test('all optional settings', () => {
            const target = join(testDir, 'all-settings');
            fs.write(join(target, 'webapp/manifest.json'), testAppManifest);
            generate(
                target,
                {
                    ...minimalInput,
                    settings: {
                        enhanceI18n: true,
                        variantManagement: 'Page'
                    }
                },
                fs
            );

            expect(fs.readJSON(join(target, 'webapp/manifest.json'))).toMatchSnapshot();
        });

        test('simple inbound navigation', () => {
            const target = join(testDir, 'with-nav');
            fs.write(join(target, 'webapp/manifest.json'), testAppManifest);
            generate(
                target,
                {
                    ...minimalInput,
                    navigation: {
                        sourcePage: 'RootEntityListReport',
                        navEntity: minimalInput.entity,
                        navKey: true
                    }
                },
                fs
            );
            expect((fs.readJSON(join(target, 'webapp/manifest.json')) as any)?.['sap.ui5'].routing).toMatchSnapshot();
        });

        test('simple nested navigation', () => {
            const target = join(testDir, 'with-nested-nav');
            fs.write(join(target, 'webapp/manifest.json'), testAppManifest);
            generate(
                target,
                {
                    ...minimalInput,
                    navigation: {
                        sourcePage: 'RootEntityObjectPage',
                        navEntity: `to_${minimalInput.entity}`,
                        navKey: true
                    }
                },
                fs
            );
            expect((fs.readJSON(join(target, 'webapp/manifest.json')) as any)?.['sap.ui5'].routing).toMatchSnapshot();
        });
    });
});
