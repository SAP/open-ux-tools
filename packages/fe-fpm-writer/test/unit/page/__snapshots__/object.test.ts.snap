// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectPage generate all optional settings 1`] = `
Object {
  "sap.app": Object {
    "id": "my.test.App",
  },
  "sap.ui5": Object {
    "dependencies": Object {
      "libs": Object {
        "sap.fe.templates": Object {},
      },
    },
    "routing": Object {
      "routes": Array [
        Object {
          "name": "RootEntityListReport",
          "pattern": ":?query:",
          "target": "TestListReport",
        },
        Object {
          "name": "RootEntityObjectPage",
          "pattern": "RootEntity({RootEntityKey}):?query:",
          "target": "RootEntityObjectPage",
        },
        Object {
          "name": "OtherEntityObjectPage",
          "pattern": "OtherEntity:?query:",
          "target": "OtherEntityObjectPage",
        },
      ],
      "targets": Object {
        "OtherEntityObjectPage": Object {
          "id": "OtherEntityObjectPage",
          "name": "sap.fe.templates.ObjectPage",
          "options": Object {
            "settings": Object {
              "enhanceI18n": "i18n/customOtherEntityObjectPage.properties",
              "entitySet": "OtherEntity",
              "navigation": Object {},
              "variantManagement": "Page",
            },
          },
          "type": "Component",
        },
        "RootEntityListReport": Object {},
        "RootEntityObjectPage": Object {},
      },
    },
  },
}
`;

exports[`ObjectPage generate minimal input 1`] = `
Object {
  "sap.app": Object {
    "id": "my.test.App",
  },
  "sap.ui5": Object {
    "dependencies": Object {
      "libs": Object {
        "sap.fe.templates": Object {},
      },
    },
    "routing": Object {
      "routes": Array [
        Object {
          "name": "RootEntityListReport",
          "pattern": ":?query:",
          "target": "TestListReport",
        },
        Object {
          "name": "RootEntityObjectPage",
          "pattern": "RootEntity({RootEntityKey}):?query:",
          "target": "RootEntityObjectPage",
        },
        Object {
          "name": "OtherEntityObjectPage",
          "pattern": "OtherEntity:?query:",
          "target": "OtherEntityObjectPage",
        },
      ],
      "targets": Object {
        "OtherEntityObjectPage": Object {
          "id": "OtherEntityObjectPage",
          "name": "sap.fe.templates.ObjectPage",
          "options": Object {
            "settings": Object {
              "entitySet": "OtherEntity",
              "navigation": Object {},
            },
          },
          "type": "Component",
        },
        "RootEntityListReport": Object {},
        "RootEntityObjectPage": Object {},
      },
    },
  },
}
`;

exports[`ObjectPage generate simple inbound navigation 1`] = `
Object {
  "routes": Array [
    Object {
      "name": "RootEntityListReport",
      "pattern": ":?query:",
      "target": "TestListReport",
    },
    Object {
      "name": "RootEntityObjectPage",
      "pattern": "RootEntity({RootEntityKey}):?query:",
      "target": "RootEntityObjectPage",
    },
    Object {
      "name": "OtherEntityObjectPage",
      "pattern": "OtherEntity({OtherEntityKey}):?query:",
      "target": "OtherEntityObjectPage",
    },
  ],
  "targets": Object {
    "OtherEntityObjectPage": Object {
      "id": "OtherEntityObjectPage",
      "name": "sap.fe.templates.ObjectPage",
      "options": Object {
        "settings": Object {
          "entitySet": "OtherEntity",
          "navigation": Object {},
        },
      },
      "type": "Component",
    },
    "RootEntityListReport": Object {
      "options": Object {
        "settings": Object {
          "navigation": Object {
            "OtherEntity": Object {
              "detail": Object {
                "route": "OtherEntityObjectPage",
              },
            },
          },
        },
      },
    },
    "RootEntityObjectPage": Object {},
  },
}
`;

exports[`ObjectPage generate simple nested navigation 1`] = `
Object {
  "routes": Array [
    Object {
      "name": "RootEntityListReport",
      "pattern": ":?query:",
      "target": "TestListReport",
    },
    Object {
      "name": "RootEntityObjectPage",
      "pattern": "RootEntity({RootEntityKey}):?query:",
      "target": "RootEntityObjectPage",
    },
    Object {
      "name": "OtherEntityObjectPage",
      "pattern": "RootEntity({RootEntityKey})/to_OtherEntity({to_OtherEntityKey}):?query:",
      "target": "OtherEntityObjectPage",
    },
  ],
  "targets": Object {
    "OtherEntityObjectPage": Object {
      "id": "OtherEntityObjectPage",
      "name": "sap.fe.templates.ObjectPage",
      "options": Object {
        "settings": Object {
          "entitySet": "OtherEntity",
          "navigation": Object {},
        },
      },
      "type": "Component",
    },
    "RootEntityListReport": Object {},
    "RootEntityObjectPage": Object {
      "options": Object {
        "settings": Object {
          "navigation": Object {
            "to_OtherEntity": Object {
              "detail": Object {
                "route": "OtherEntityObjectPage",
              },
            },
          },
        },
      },
    },
  },
}
`;
